### PR TITLE
add command to infer pinot schema from json data

### PR DESCRIPTION
**Pull request by yupeng9**
## Description
Part of https://github.com/apache/incubator-pinot/issues/6904

Add a command to infer the Pinot schema from the JSON data, with the complex type handling.

For example, a run of 
```
JsonToPinotSchema -timeColumnName hoursSinceEpoch -jsonFile  //tmp/test/test.json -pinotSchemaName json-schema -outputDir /tmp/test -unnestFields=payload.commits 
```

On the Github event data
```
{
	"id": "7044874109",
	"type": "PushEvent",
	"actor": {
		"id": 18542751,
		"login": "LimeVista",
		"display_login": "LimeVista",
		"gravatar_id": "",
		"url": "https://api.github.com/users/LimeVista",
		"avatar_url": "https://avatars.githubusercontent.com/u/18542751?"
	},
	"repo": {
		"id": 115911530,
		"name": "LimeVista/Tapes",
		"url": "https://api.github.com/repos/LimeVista/Tapes"
	},
	"payload": {
		"push_id": 2226018068,
		"size": 1,
		"distinct_size": 1,
		"ref": "refs/heads/master",
		"head": "c5fc8b32a9ead1eba315d97410cb4ac1e6ca1774",
		"before": "892d872c5d3f24cc6837900c9f4618dc2fe92930",
		"commits": [{
			"sha": "c5fc8b32a9ead1eba315d97410cb4ac1e6ca1774",
			"author": {
				"name": "Lime",
				"email": "4cc153d999e24274955157fc813e6f92f821525d@outlook.com"
			},
			"message": "Merge branch 'master' of https://github.com/LimeVista/Tapes\\n\\n# Conflicts:\\n#\\t.gitignore",
			"distinct": true,
			"url": "https://api.github.com/repos/LimeVista/Tapes/commits/c5fc8b32a9ead1eba315d97410cb4ac1e6ca1774"
		}]
	},
	"public": true,
	"created_at": "2018-01-01T11:00:00Z"
}
```

Will output
```
{
  "schemaName" : "json-schema",
  "dimensionFieldSpecs" : [ {
    "name" : "id",
    "dataType" : "STRING"
  }, {
    "name" : "type",
    "dataType" : "STRING"
  }, {
    "name" : "actor.id",
    "dataType" : "INT"
  }, {
    "name" : "actor.login",
    "dataType" : "STRING"
  }, {
    "name" : "actor.display_login",
    "dataType" : "STRING"
  }, {
    "name" : "actor.gravatar_id",
    "dataType" : "STRING"
  }, {
    "name" : "actor.url",
    "dataType" : "STRING"
  }, {
    "name" : "actor.avatar_url",
    "dataType" : "STRING"
  }, {
    "name" : "repo.id",
    "dataType" : "INT"
  }, {
    "name" : "repo.name",
    "dataType" : "STRING"
  }, {
    "name" : "repo.url",
    "dataType" : "STRING"
  }, {
    "name" : "payload.push_id",
    "dataType" : "LONG"
  }, {
    "name" : "payload.size",
    "dataType" : "INT"
  }, {
    "name" : "payload.distinct_size",
    "dataType" : "INT"
  }, {
    "name" : "payload.ref",
    "dataType" : "STRING"
  }, {
    "name" : "payload.head",
    "dataType" : "STRING"
  }, {
    "name" : "payload.before",
    "dataType" : "STRING"
  }, {
    "name" : "payload.commits.sha",
    "dataType" : "STRING"
  }, {
    "name" : "payload.commits.author.name",
    "dataType" : "STRING"
  }, {
    "name" : "payload.commits.author.email",
    "dataType" : "STRING"
  }, {
    "name" : "payload.commits.message",
    "dataType" : "STRING"
  }, {
    "name" : "payload.commits.distinct",
    "dataType" : "BOOLEAN"
  }, {
    "name" : "payload.commits.url",
    "dataType" : "STRING"
  }, {
    "name" : "public",
    "dataType" : "BOOLEAN"
  }, {
    "name" : "created_at",
    "dataType" : "STRING"
  } ]
}%
```

### CODE REVIEW

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  @Option(name = "-pinotSchemaName", required = true, metaVar = "<string>", usage = "Pinot schema name")
+  String _pinotSchemaName;
+
+  @Option(name = "-dimensions", metaVar = "<string>", usage = "Comma separated dimension column names.")
+  String _dimensions;
+
+  @Option(name = "-metrics", metaVar = "<string>", usage = "Comma separated metric column names.")
+  String _metrics;
+
+  @Option(name = "-timeColumnName", metaVar = "<string>", usage = "Name of the time column.")
```
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java:43]: We already deprecated the TIME field, change it to `-dateTimeColumns`?

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java [hunk 2]
```diff
@@ -380,4 +389,120 @@ private static void unnestResults(List<Map<String, String>> currentResults,
...
+      Preconditions.checkNotNull(fieldType, "Field type not specified for field: %s", name);
+      switch (fieldType) {
+        case DIMENSION:
+          pinotSchema.addField(new DimensionFieldSpec(name, dataType, isSingleValueField));
+          break;
+        case METRIC:
+          Preconditions.checkState(isSingleValueField, "Metric field: %s cannot be multi-valued", name);
+          pinotSchema.addField(new MetricFieldSpec(name, dataType));
+          break;
+        case TIME:
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java:491]: We already deprecated TIME field. Let's only support `DATE_TIME` here

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java [hunk 3]
```diff
@@ -380,4 +389,120 @@ private static void unnestResults(List<Map<String, String>> currentResults,
       unnestResults(newCurrentResults, nestedResultsList, index + 1, nonNestedResult, outputResults);
     }
   }
+
+  public static Schema getPinotSchemaFromJsonFile(File jsonFile,
+      @Nullable Map<String, FieldSpec.FieldType> fieldTypeMap, @Nullable TimeUnit timeUnit, List<String> unnestFields,
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java:394]: unnestFields should also be nullable

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java [hunk 4]
```diff
@@ -380,4 +389,120 @@ private static void unnestResults(List<Map<String, String>> currentResults,
...
+    }
+    return pinotSchema;
+  }
+
+  private static void inferPinotSchemaFromJsonNode(JsonNode jsonNode, Schema pinotSchema, String path,
+      @Nullable Map<String, FieldSpec.FieldType> fieldTypeMap, @Nullable TimeUnit timeUnit, List<String> unnestFields,
+      String delimiter) {
+    if (jsonNode.isNull()) {
+      // do nothing
+      return;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java:421]: Should we throw exception when schema cannot be inferred?

Comment (Jackie-Jiang) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java:424]: Either way is fine, just the null field won't be added to the pinot schema

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        fieldTypes.put(column, FieldSpec.FieldType.DIMENSION);
+      }
+    }
+    if (_metrics != null) {
+      for (String column : _metrics.split("\\s*,\\s*")) {
+        fieldTypes.put(column, FieldSpec.FieldType.METRIC);
+      }
+    }
+    if (_dateTimeColumnName != null) {
+      fieldTypes.put(_dateTimeColumnName, FieldSpec.FieldType.TIME);
```
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java:143]:
```suggestion
      fieldTypes.put(_dateTimeColumnName, FieldSpec.FieldType.DATE_TIME);
```

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    return fieldTypes;
+  }
+
+  private List<String> buildUnnestFields() {
+    List<String> unnestFields = new ArrayList<>();
+    if (_unnestFields != null) {
+      for (String field : _unnestFields.split(",")) {
```
Comment (Jackie-Jiang) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/command/JsonToPinotSchema.java:151]: You might want to remove white spaces?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java [hunk 7]
```diff
@@ -380,4 +389,120 @@ private static void unnestResults(List<Map<String, String>> currentResults,
       unnestResults(newCurrentResults, nestedResultsList, index + 1, nonNestedResult, outputResults);
     }
   }
+
+  public static Schema getPinotSchemaFromJsonFile(File jsonFile,
```
[the remark this answers is not in the document]
Comment (yupeng9) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java:412]: done

**Pull request merged**